tests/test_url.py

```diff
@@ -1657,6 +1657,83 @@ def test_join_from_rfc_3986_abnormal(url, expected):
     assert base.join(url) == expected
 
 
+EMPTY_SEGMENTS = [
+    (
+        "https://web.archive.org/web/",
+        "./https://github.com/aio-libs/yarl",
+        "https://web.archive.org/web/https://github.com/aio-libs/yarl",
+    ),
+    (
+        "https://web.archive.org/web/https://github.com/",
+        "aio-libs/yarl",
+        "https://web.archive.org/web/https://github.com/aio-libs/yarl",
+    ),
+]
+
+
+@pytest.mark.parametrize("base,url,expected", EMPTY_SEGMENTS)
+def test_join_empty_segments(base, url, expected):
+    base = URL(base)
+    url = URL(url)
+    expected = URL(expected)
+    joined = base.join(url)
+    assert joined == expected
+
+
+SIMPLE_BASE = "http://a/b/c/d"
+URLLIB_URLJOIN = [
+    ("", "http://a/b/c/g?y/./x", "http://a/b/c/g?y/./x"),
+    ("", "http://a/./g", "http://a/./g"),
+    ("svn://pathtorepo/dir1", "dir2", "svn://pathtorepo/dir2"),
+    ("svn+ssh://pathtorepo/dir1", "dir2", "svn+ssh://pathtorepo/dir2"),
+    ("ws://a/b", "g", "ws://a/g"),
+    ("wss://a/b", "g", "wss://a/g"),
+    # test for issue22118 duplicate slashes
+    (SIMPLE_BASE + "/", "foo", SIMPLE_BASE + "/foo"),
+    # Non-RFC-defined tests, covering variations of base and trailing
+    # slashes
+    ("http://a/b/c/d/e/", "../../f/g/", "http://a/b/c/f/g/"),
+    ("http://a/b/c/d/e", "../../f/g/", "http://a/b/f/g/"),
+    ("http://a/b/c/d/e/", "/../../f/g/", "http://a/f/g/"),
+    ("http://a/b/c/d/e", "/../../f/g/", "http://a/f/g/"),
+    ("http://a/b/c/d/e/", "../../f/g", "http://a/b/c/f/g"),
+    ("http://a/b/", "../../f/g/", "http://a/f/g/"),
+    ("a", "b", "b"),
+]
+
+
+@pytest.mark.parametrize("base,url,expected", URLLIB_URLJOIN)
+def test_join_cpython_urljoin(base, url, expected):
+    # tests from cpython urljoin
+    base = URL(base)
+    url = URL(url)
+    expected = URL(expected)
+    joined = base.join(url)
+    assert joined == expected
+
+
+URLLIB_URLJOIN_FAIL = [
+    (
+        "http:///",
+        "..",
+        "http:///",
+        TypeError,
+        "unsupported operand type\\(s\\) for \\+: 'NoneType' and 'str'",
+    ),
+]
+
+
+@pytest.mark.xfail(raises=TypeError, strict=True)
+@pytest.mark.parametrize("base,url,expected,exc,match", URLLIB_URLJOIN_FAIL)
+def test_join_cpython_urljoin_fail(base, url, expected, exc, match):
+    # non portable tests from cpython urljoin
+    base = URL(base)
+    url = URL(url)
+    with pytest.raises(exc, match=match) as e:
+        base.join(url)
+    raise e.value
```

Comment:
Why have a `raise` statement in tests? What does the test check, then?

Reply:
If an unexpected exception is raised (or none at all), `pytest.raises()` would display it already.

Reply:
pytest.raises() is used to match the exceptions message, re-raised to have a xfail.

Reply:
I'm not sure if there's some confusion over how xfail works here.

xfail means the test will fail because something in our code is broken, and we should fix it later. Whereas this test just seems to fail for no reason.

pytest.raises() will fail a test if an exception is not raised. So, if the expected case is that exceptions are raised, then removing the `raise` and removing the `xfail` would give us a correct, working test.

Reply:
I reopened this thread since it wasn't addressed. It sounds like this might not even need to be an xfailing test.

@commonism could you explain more verbosely what's the intent and design behind this specific test?

Reply:
xfail does not allow to specify a match for the exception string, pytest.raises does.
Combining both allows to create a tight testing scenario for broken behavior, and indicate this current behavior is considered broken and would be great to have addressed `later`.

I hope this explains my intention - I just wanted to include the match for the exception string in the test.

Reply:
It looks like this test is written to verify that the code produces the wrong result.

Test should be written to indicate what we expect the code should do, and than marked with xfail to indicate the code currently does not work as expected.

Reply:
> xfail does not allow to specify a match for the exception string, pytest.raises does.

So what's happening here, is that the test is expecting the code to not raise an exception.
We then catch the exception with pytest.raises() in order to use the match argument.
Then we reraise that exception in order to turn it back into an xfail.

This is causing so much confusion, I'd vote for just doing without the match argument and remove the pytest.raises().

Reply:
Or possibly, just replace the `raise` with `pytest.xfail()` and xfail it in the test code instead.